cmd/http-trace.go

```diff
@@ -0,0 +1,212 @@
+	fmt.Fprintf(w, "%d %s\n", statusCode, http.StatusText(statusCode))
+	for k, v := range headers {
+		fmt.Fprintf(w, "%s: %s\n", k, v[0])
+	}
+}
+
+func (r *recordResponseWriter) WriteHeader(i int) {
+	r.statusCode = i
+	if !r.headersLogged {
+		writeHeaders(&r.headers, i, r.ResponseWriter.Header())
+		r.headersLogged = true
+	}
+	r.ResponseWriter.WriteHeader(i)
+}
+
+func (r *recordResponseWriter) Write(p []byte) (n int, err error) {
+	n, err = r.ResponseWriter.Write(p)
+	// We log after calling ResponseWriter.Write() because this latter
+	// proactively prepares headers when WriteHeader is not called yet.
+	if !r.headersLogged {
+		writeHeaders(&r.headers, http.StatusOK, r.ResponseWriter.Header())
+		r.headersLogged = true
+	}
+	if r.statusCode != http.StatusOK && r.statusCode != http.StatusNoContent && r.statusCode != 0 {
+		// We always log error responses.
+		r.body.Write(p)
+		return
+	}
+	if r.logBody {
+		r.body.Write(p)
+		return
+	}
+	return n, err
+}
+
+func (r *recordResponseWriter) Flush() {
+	r.ResponseWriter.(http.Flusher).Flush()
+}
+
+// Return response headers.
+func (r *recordResponseWriter) Headers() []byte {
+	return r.headers.Bytes()
+}
+
+// Return response body.
+func (r *recordResponseWriter) Body() []byte {
+	return r.body.Bytes()
+}
+
+// Log headers and body.
+func httpTracelogAll(f http.HandlerFunc) http.HandlerFunc {
+	if !globalHTTPTrace {
+		return f
+	}
+	return httpTracelog(f, true)
+}
+
+// Log only the headers.
+func httpTracelogHeaders(f http.HandlerFunc) http.HandlerFunc {
+	if !globalHTTPTrace {
+		return f
+	}
+	return httpTracelog(f, false)
+}
+
+func httpTracelog(f http.HandlerFunc, logBody bool) http.HandlerFunc {
+	name := runtime.FuncForPC(reflect.ValueOf(f).Pointer()).Name()
+	name = strings.TrimPrefix(name, "github.com/minio/minio/cmd.")
+	bodyPlaceHolder := []byte("<BODY>")
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		const timeFormat = "2006-01-02 15:04:05 -0700"
+		var reqBodyRecorder *recordRequest
+
+		// Generate short random request ID
+		reqID := fmt.Sprintf("%f", float64(time.Now().UnixNano())/1e10)
+
+		reqBodyRecorder = &recordRequest{Reader: r.Body, logBody: logBody}
+		r.Body = ioutil.NopCloser(reqBodyRecorder)
+
+		// Setup a http response body recorder
+		respBodyRecorder := &recordResponseWriter{ResponseWriter: w, logBody: logBody}
+
+		b := bytes.NewBuffer(nil)
+		fmt.Fprintf(b, "[REQUEST %s] [%s] [%s]\n", name, reqID, time.Now().Format(timeFormat))
+
+		f(respBodyRecorder, r)
+
+		// Build request log and write it to log file
+		fmt.Fprintf(b, "%s %s", r.Method, r.URL.Path)
+		if r.URL.RawQuery != "" {
+			fmt.Fprintf(b, "?%s", r.URL.RawQuery)
+		}
+		fmt.Fprintf(b, "\n")
+
+		fmt.Fprintf(b, "Host: %s\n", r.Host)
+		for k, v := range r.Header {
+			fmt.Fprintf(b, "%s: %s\n", k, v[0])
+		}
+		fmt.Fprintf(b, "\n")
+		if logBody {
+			bodyContents := reqBodyRecorder.Data()
+			if bodyContents != nil {
+				b.Write(bodyContents)
+				fmt.Fprintf(b, "\n")
```

Comment:
We can have some log, generated concurrently by parallel request handling, overlapped. We can add a mutex here. Or we can append '\n' to bodyContents since b.Write() is atomic, as I tested last time.

Reply:
no, actually it is printing to a buffer `b` and in the end it logs to Stdout in one go so that logs are not interspersed

Reply:
> no, actually it is printing to a buffer b and in the end it logs to Stdout in one go so that logs are not interspersed

oups, yes.